astra/src/main/java/com/slack/astra/bulkIngestApi/BulkIngestS3Producer.java

```diff
@@ -0,0 +1,344 @@
+
+  // Metric name constants
+  public static final String S3_UPLOAD_COUNTER = "bulk_ingest_producer_s3_wal_uploads_total";
+  public static final String S3_SPANS_UPLOADED_COUNTER =
+      "bulk_ingest_producer_s3_wal_spans_uploaded_total";
+  public static final String S3_UPLOAD_TIMER = "bulk_ingest_producer_s3_wal_upload_duration";
+  public static final String S3_BYTES_UPLOADED_COUNTER =
+      "bulk_ingest_producer_s3_wal_bytes_uploaded_total";
+  public static final String S3_UPLOAD_FAILURES_COUNTER =
+      "bulk_ingest_producer_s3_wal_upload_failures_total";
+  public static final String KAFKA_POINTER_FAILURES_COUNTER =
+      "bulk_ingest_producer_s3_wal_kafka_failures_total";
+  public static final String STOP_INGESTION_COUNTER =
+      "bulk_ingest_producer_s3_wal_stop_ingestion_total";
+  private final BlobStore blobStore;
+  protected final String walBucket;
+  private final Timer s3UploadTimer;
+  private final int maxBufferTimeMs;
+  private final int maxRequestsPerBatch;
+  private final int minBatchSize;
+  private final int bufferWaitMs;
+
+  public BulkIngestS3Producer(
+      final DatasetMetadataStore datasetMetadataStore,
+      final AstraConfigs.PreprocessorConfig preprocessorConfig,
+      final MeterRegistry meterRegistry,
+      BlobStore blobStore) {
+
+    super(datasetMetadataStore, preprocessorConfig, meterRegistry);
+
+    // Initialize S3Producer specific fields
+    AstraConfigs.S3WalBufferConfig s3WalBufferConfig =
+        preprocessorConfig.getS3WalConfig().getBufferConfig();
+    this.blobStore = blobStore;
+    this.walBucket = preprocessorConfig.getS3WalConfig().getS3Config().getS3Bucket();
+    this.kafkaTopic = preprocessorConfig.getKafkaConfig().getKafkaTopic();
+    this.s3UploadTimer = meterRegistry.timer(S3_UPLOAD_TIMER);
+    this.producerSleepMs =
+        Integer.parseInt(System.getProperty("astra.bulkIngest.S3producerSleepMs", "100"));
+    this.maxBufferTimeMs = s3WalBufferConfig.getMaxBufferTimeMs();
+    this.maxRequestsPerBatch = s3WalBufferConfig.getMaxRequestsPerBatch();
+    this.minBatchSize = Math.max(1, maxRequestsPerBatch / 4);
+    this.bufferWaitMs = Math.max(10, maxBufferTimeMs / 10);
+  }
+
+  @Override
+  protected void run() throws Exception {
+
+    long lastProcessTime = System.currentTimeMillis();
+
+    while (isRunning()) {
+
+      // Check queue size first, don't drain if not ready
+      int availableRequests = pendingRequests.size();
+      if (availableRequests == 0) {
+        // No requests available, sleep and continue
+        try {
+          stallCounter.increment();
+          Thread.sleep(producerSleepMs);
+        } catch (InterruptedException e) {
+          return;
+        }
+        continue;
+      }
+
+      // Check if we should wait for more requests
+      if (shouldWaitForMoreRequests(availableRequests, lastProcessTime)) {
+        // Don't drain anything, just sleep and wait for more
+        Thread.sleep(bufferWaitMs);
```

Comment:
you could do the max time / 10 calc here, which would make it more clear what this is doing, without adding much overhead. Or, you could reuse the producerSleepMs, which close enough in semantics and can be separately tuned if we needed to do that.

I'm not sure if we could collapse the conditionals too though, because I'm not sure if waiting for more requests to fill into the queue is equivalent to waiting on an empty queue.